### DISCUSSION TITLE
I need to press tab.

### DISCUSSION THREAD

**Discussion by kfsone** [community/community#17238] [Copilot Conversations]
Synopsis:
Because copilot can produce suggestions at whitespace and because it can produce them faster than the key-repeat/typing speed of a human, it surfaces issues with IDE input mechanisms that aren't usually apparent to the user. In particular, it can generate a suggestion within two <tab> keystrokes causing input-fighting.

Description:
We have a whitespace-OCD codebase where things need to be lined up, which is antithetical to many modern BCPs and style guides, but it's what I have to work in.

The result is an incredibly unpleasant experience working with co-pilot because it becomes virtually impossible to add indentation by pressing tab.

Specifically: I press tab and a suggestion I never saw gets inserted.

I could accept that it's simply happening so quickly that I haven't registered it, but if that were the case then pressing _esc,tab_  or  _tab,ctrl+z,esc,tab_ should work.

They don't.

I've tried waiting 1 and 2 beats longer before pressing esc, doesn't work. If I press esc before the suggestion is visible (and I mean as in the input needs to be processed on or after the screen refresh where the suggestion is rendered) then pressing tab will insert the suggestion.

![CopilotTab](https://user-images.githubusercontent.com/323009/170110836-80a49bdb-4ef0-41bf-8589-aa0d003b432d.gif)

Obviously this only affects tab, perhaps it would make sense to allow a delay after pressing tab in the 280-350ms range.

It should be possible to press tab twice quickly in the editor without that equating to "tab + suggestion".

-Oliver

**Comment by luketych** [comment]
I second this

**Comment by cmp-nct** [comment]
Related to https://github.com/community/community/discussions/36266
It's not just indentations but in general the configured code style that needs to be adhered.
Given how smart the style formatters are today it's probably not too hard to adapt the plugins, they basically just need to run the completed code through the IDE's built in formatting tool. If that's possible in a plugin.

**Comment by felipelalli** [comment]
Related? #9978

**Reply by kash831** [reply]
Why does gut hub make everything so damn difficult 

- [x

**Reply by luketych** [reply]
**Gut** Hub
